controller/formation.go

```diff
@@ -281,29 +316,31 @@ func (r *FormationRepo) sendUpdatedSince(ch chan *ct.ExpandedFormation, stopCh <
 		if err != nil {
 			return err
 		}
-		select {
-		case ch <- ef:
-		case <-stopCh:
+		if ok := sub.notify(ef); !ok {
 			return nil
 		}
 	}
-	ch <- &ct.ExpandedFormation{} // sentinel
+	sub.notify(&ct.ExpandedFormation{}) // sentinel
 	return rows.Err()
 }
 
-func (r *FormationRepo) Unsubscribe(ch chan *ct.ExpandedFormation) {
+func (r *FormationRepo) Unsubscribe(sub *FormationSubscription) {
 	r.subMtx.Lock()
 	defer r.subMtx.Unlock()
-	r.unsubscribeLocked(ch)
+	r.unsubscribeLocked(sub)
 }
 
-func (r *FormationRepo) unsubscribeLocked(ch chan *ct.ExpandedFormation) {
-	go func() {
+func (r *FormationRepo) unsubscribeLocked(sub *FormationSubscription) {
+	go func(ch chan *ct.ExpandedFormation) {
 		// drain to prevent deadlock while removing the listener
 		for range ch {
 		}
-	}()
-	delete(r.subscriptions, ch)
+	}(sub.ch)
```

Comment:
Actually it matters since `sub.ch` will become `nil` when the subscription is closed.

[This deadlocks](https://play.golang.org/p/3MRVr6j___).

[This does not dealock](https://play.golang.org/p/GADxvT3LZ5).